File: C++/stack.cpp
#include "stack.h"

Stack::Stack(size_t size) : capacity(size), head(-1) {
    data = new string[capacity];
}

Stack::~Stack() { 
    delete[] data; 
}

void Stack::push(string value) {
    if (head == (int)capacity - 1) {
        throw overflow_error("Stack is full");
    }
    data[++head] = value;
}

string Stack::pop() {
    if (head == -1) {
        throw underflow_error("Stack is empty");
    }
    return data[head--];
}

string Stack::peek() {
    if (head == -1) {
        throw underflow_error("Stack is empty");
    }
    return data[head];
}

bool Stack::isEmpty() { 
    return head == -1; 
}

size_t Stack::size() { 
    return head + 1; 
}

vector<string> Stack::toVector() const {
    vector<string> result;
    for (int i = 0; i <= head; ++i) {
        result.push_back(data[i]);
    }
    return result;
}

void Stack::fromVector(const vector<string>& elements) {
    delete[] data;
    capacity = elements.size() + 10;
    data = new string[capacity];
    head = -1;
    for (const auto& element : elements) {
        push(element);
    }
}

File: GO/stack.go
package main

import "errors"

type Stack struct {
	data     []string
	capacity int
}

func NewStack(capacity int) *Stack {
	return &Stack{
		data:     make([]string, 0, capacity),
		capacity: capacity,
	}
}

func (s *Stack) Push(value string) error {
	if len(s.data) >= s.capacity {
		return errors.New("stack is full")
	}
	s.data = append(s.data, value)
	return nil
}

func (s *Stack) Pop() (string, error) {
	if len(s.data) == 0 {
		return "", errors.New("stack is empty")
	}
	value := s.data[len(s.data)-1]
	s.data = s.data[:len(s.data)-1]
	return value, nil
}

func (s *Stack) Peek() (string, error) {
	if len(s.data) == 0 {
		return "", errors.New("stack is empty")
	}
	return s.data[len(s.data)-1], nil
}

func (s *Stack) IsEmpty() bool {
	return len(s.data) == 0
}

func (s *Stack) Size() int {
	return len(s.data)
}

func (s *Stack) ToVector() []string {
	result := make([]string, len(s.data))
	copy(result, s.data)
	return result
}

func (s *Stack) FromVector(elements []string) {
	s.data = make([]string, 0, len(elements)+10)
	s.capacity = len(elements) + 10
	for _, element := range elements {
		s.data = append(s.data, element)
	}
}
